include/picongpu/plugins/saxs/Saxs.hpp

```diff
@@ -0,0 +1,815 @@
+        {
+            using namespace mpi;
+            // No action is necessary on the first step.
+            if ( step == 0u ) return;
+            // Copy data calculated on GPU , on last step, to CPU memory.
+            amplitude->deviceToHost( );
+            // Avoid deadlock between not finished pmacc tasks and mpi blocking
+            // collectives.
+            __getTransactionEvent( ).waitForFinished( );
+            // MPI asynchronous send & receive:
+            int bytesToSend = sizeof( complex_X ) / sizeof( char );
+            bytesToSend *= bufferSize_i;
+
+            // An mpi request to monitor a non blocking send transaction.
+            MPI_Request transactionRequest;
+            // Pass data to the next node.
+            MPI_CHECK( MPI_Isend(
+                amplitude->getHostBuffer( ).getBasePointer( ),
+                bytesToSend,
+                MPI_BYTE,
+                ( mpiRank + 1 ) % countRanks,
+                0,
+                gc.getCommunicator( ).getMPIComm( ),
+                & transactionRequest
+            ) );
+            // Receive from the proceeding node (blocking transaction).
+            int receiveFrom = ( mpiRank == 0u ) ? countRanks - 1 : mpiRank - 1;
+            MPI_CHECK( MPI_Recv(
+                amplitudeReceive.data( ),
+                bytesToSend,
+                MPI_BYTE,
+                std::move( receiveFrom ),
+                0,
+                gc.getCommunicator( ).getMPIComm( ),
+                MPI_STATUS_IGNORE
+            ) );
+
+            // Wait for the send transaction to end.
+            MPI_Wait(
+                & transactionRequest,
+                MPI_STATUS_IGNORE
+            );
+            // Copy the received data to the host buffer.
+            copyVectorToBuffer(
+                amplitudeReceive,
+                amplitude->getHostBuffer( )
+            );
+            // Copy the received data to the device so it can be used as
+            // output in this step.
+            amplitude->hostToDevice( );
+            __getTransactionEvent( ).waitForFinished( );
+        }
+
+
+        /** Calculates a form factor number density of the species.
+         *
+         * @param dc data connector
+         * @param globalOffset offset from the global to the local domain
+         * @return data box containing the calculated data.
+         */
+        HINLINE FieldTmp::DataBoxType calculateDensity(
+            DataConnector & dc,
+            DataSpace< simDim > & globalOffset
+        )
+        {
+            // Check if there is at least one unused field available.
+            PMACC_CASSERT_MSG(
+                _please_allocate_at_least_one_FieldTmp_in_memory_param,
+                fieldTmpNumSlots > 0
+            );
+            // Get a field for density storage.
+            auto tmpField = dc.get< FieldTmp >(
+                FieldTmp::getUniqueId( 0 ),
+                true
+            );
```

Comment:
It seems this is not released in this method. Is it on purpose?

Reply:
Well yes, the field is released at the end of `notify`. The values are used by the lernel later on so I don't think that I can release this already here :)

Reply:
You are right, releasing is done at the right moment. I got confused because it did not happen in the same scope as acquiring the field, which makes it harder to follow. However, I do not think changes are needed here.